File: Practica 2 - Java RMI/GestorBibliotecaClient/src/gestor/biblioteca/client/menu/MainMenu.java
package gestor.biblioteca.client.menu;

import gestor.biblioteca.client.properties.GestorBibliotecaUserProperties;
import gestor.biblioteca.client.utils.TDatosRepositorioUtils;
import gestor.biblioteca.service.GestorBibliotecaIntf;
import gestor.biblioteca.service.models.TDatosRepositorio;
import gestor.biblioteca.service.models.TLibro;
import gestor.biblioteca.service.models.handling.BookSearcher;
import static gestor.biblioteca.service.models.handling.BookSearcher.ALL_SEARCH_FIELD;
import static gestor.biblioteca.service.models.handling.BookSearcher.AUTHOR_SEARCH_FIELD;
import static gestor.biblioteca.service.models.handling.BookSearcher.COUNTRY_SEARCH_FIELD;
import static gestor.biblioteca.service.models.handling.BookSearcher.ISBN_SEARCH_FIELD;
import static gestor.biblioteca.service.models.handling.BookSearcher.LANGUAGE_SEARCH_FIELD;
import static gestor.biblioteca.service.models.handling.BookSearcher.TITLE_SEARCH_FIELD;
import gestor.biblioteca.service.util.BookUtils;
import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Saúl Rodríguez Naranjo
 */
public class MainMenu
{

    public static final String MAIN_MENU_TITLE = "GESTOR BIBLIOTECARIO 2.0 (M.PRINCIPAL)";
    public static final String OPTION_0_TITLE = "0.- Salir";
    public static final String OPTION_1_TITLE = "1.- M. Administración";
    public static final String OPTION_2_TITLE = "2.- Consulta de libros";
    public static final String OPTION_3_TITLE = "3.- Préstamo de libros";
    public static final String OPTION_4_TITLE = "4.- Devolución de libros";

    private GestorBibliotecaIntf gestorBiblioteca;
    private Scanner scanner;

    public MainMenu(GestorBibliotecaIntf gestorBiblioteca, Scanner scanner)
    {
        this.gestorBiblioteca = gestorBiblioteca;
        this.scanner = scanner;
    }

    public void showMainMenu()
    {
        System.out.println(MAIN_MENU_TITLE);
        System.out.println("*******************************");
        System.out.println(OPTION_1_TITLE);
        System.out.println(OPTION_2_TITLE);
        System.out.println(OPTION_3_TITLE);
        System.out.println(OPTION_4_TITLE);
        System.out.println(OPTION_0_TITLE);
    }

    public void executeOption0()
    {
        System.exit(0);
    }

    public void executeOption1()
    {
        try
        {
            //Clean buffer
            scanner.nextLine();

            String password = "";

            System.out.println("Por favor inserte la contraseña de administración:");
            password = scanner.nextLine();

            int connectionResult = gestorBiblioteca.Conexion(password);

            switch (connectionResult)
            {
                case -1:
                    System.out.println("ERROR: Ya hay un usuario identificado como administrador");
                    break;
                case -2:
                    System.out.println("ERROR: Contraseña errónea");
                default:
                    GestorBibliotecaUserProperties.getInstance().setAdminId(connectionResult);

                    AdministrationMenu adminMenu = new AdministrationMenu(gestorBiblioteca, scanner);

                    int option = -1;

                    do
                    {
                        adminMenu.show();

                        System.out.println("Selecciona una opción:");
                        option = scanner.nextInt();

                        adminMenu.executeOption(option);

                    } while (option != 0);
            }
        } catch (RemoteException ex)
        {
            System.out.println("Error de conexión: " + ex.getMessage());
        }
    }

    public void executeOption2()
    {
        //Clean buffer
        scanner.nextLine();

        String searchText = "";

        System.out.println("Introduce el texto a buscar: ");
        searchText = scanner.nextLine();

        String searchCode = "";

        System.out.println("Código de Búsqueda");
        System.out.println("I.- Por ISBN");
        System.out.println("T.- Por Título");
        System.out.println("A.- Por Autor");
        System.out.println("P.- Por País");
        System.out.println("D.- Por Idioma");
        System.out.println("*.- Por todos los campos");

        do
        {
            System.out.println("Introduce código: ");
            searchCode = scanner.nextLine();

            if (!searchCode.equalsIgnoreCase("i") && !searchCode.equalsIgnoreCase("t")
                    && !searchCode.equalsIgnoreCase("a") && !searchCode.equalsIgnoreCase("p")
                    && !searchCode.equalsIgnoreCase("d") && !searchCode.equalsIgnoreCase("*"))
            {
                System.out.println("ERORR: El código introducido no es válido");
            }
        } while (!searchCode.equalsIgnoreCase("i") && !searchCode.equalsIgnoreCase("t")
                && !searchCode.equalsIgnoreCase("a") && !searchCode.equalsIgnoreCase("p")
                && !searchCode.equalsIgnoreCase("d") && !searchCode.equalsIgnoreCase("*"));

        try
        {
            int totalNumberOfBooks = gestorBiblioteca.NLibros(-1);
            
            if (totalNumberOfBooks <= 0)
            {
                System.out.println("ERROR: No hay repositorios cargados en la biblioteca");

                return;
            }

            BookUtils bookUtils = new BookUtils();

            int adminId = GestorBibliotecaUserProperties.getInstance().getAdminId();

            List<TLibro> allBooks = new ArrayList<>();

            for (int i = 0; i < totalNumberOfBooks; i++)
            {
                TLibro book = gestorBiblioteca.Descargar(adminId, -1, i);

                if (book != null)
                {
                    allBooks.add(book);
                }
            }

            boolean headerShow = true;

            if (searchCode.equalsIgnoreCase(ISBN_SEARCH_FIELD))
            {
                for (int i = 0; i < allBooks.size(); i++)
                {
                    TLibro book = allBooks.get(i);

                    if (book.getIsbn().contains(searchText))
                    {
                        bookUtils.Mostrar(i, headerShow, book);
                        headerShow = false;
                    }
                }
            } else if (searchCode.equalsIgnoreCase(TITLE_SEARCH_FIELD))
            {
                for (int i = 0; i < allBooks.size(); i++)
                {
                    TLibro book = allBooks.get(i);

                    if (book.getTitulo().contains(searchText))
                    {
                        bookUtils.Mostrar(i, headerShow, book);
                        headerShow = false;
                    }
                }
            } else if (searchCode.equalsIgnoreCase(AUTHOR_SEARCH_FIELD))
            {
                for (int i = 0; i < allBooks.size(); i++)
                {
                    TLibro book = allBooks.get(i);

                    if (book.getAutor().contains(searchText))
                    {
                        bookUtils.Mostrar(i, headerShow, book);
                        headerShow = false;
                    }
                }
            } else if (searchCode.equalsIgnoreCase(COUNTRY_SEARCH_FIELD))
            {
                for (int i = 0; i < allBooks.size(); i++)
                {
                    TLibro book = allBooks.get(i);

                    if (book.getPais().contains(searchText))
                    {
                        bookUtils.Mostrar(i, headerShow, book);
                        headerShow = false;
                    }
                }
            } else if (searchCode.equalsIgnoreCase(LANGUAGE_SEARCH_FIELD))
            {
                for (int i = 0; i < allBooks.size(); i++)
                {
                    TLibro book = allBooks.get(i);

                    if (book.getIdioma().contains(searchText))
                    {
                        bookUtils.Mostrar(i, headerShow, book);
                        headerShow = false;
                    }
                }
            } else if (searchCode.equalsIgnoreCase(ALL_SEARCH_FIELD))
            {
                for (int i = 0; i < allBooks.size(); i++)
                {
                    TLibro book = allBooks.get(i);

                    if (book.getIsbn().contains(searchText) || book.getTitulo().contains(searchText)
                            || book.getAutor().contains(searchText) || book.getPais().contains(searchText)
                            || book.getIdioma().contains(searchText))
                    {
                        bookUtils.Mostrar(i, headerShow, book);
                        headerShow = false;
                    }
                }
            }

        } catch (RemoteException ex)
        {
            Logger.getLogger(AdministrationMenu.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    public void executeOption3()
    {

    }

    public void executeOption4()
    {

    }

    public void executeOption(int optionNumber)
    {
        switch (optionNumber)
        {
            case 0:
                executeOption0();
                break;
            case 1:
                executeOption1();
                break;
            case 2:
                executeOption2();
                break;
            case 3:
                executeOption3();
                break;
            case 4:
                executeOption4();
                break;
            default:
                System.out.println("opción incorrecta");
        }
    }
}

File: Practica 3 - C# Net Remoting/GestorBibliotecaService/GestorBibliotecaClient/Menu/MainMenu.cs
using GestorBibliotecaService.Data.Handling;
using GestorBibliotecaService.UserProperties;
using GestorBibliotecaService.Util;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GestorBibliotecaService.Menu
{
    public class MainMenu
    {
        public const string MAIN_MENU_TITLE = "GESTOR BIBLIOTECARIO 2.0 (M.PRINCIPAL)";
        public const string OPTION_0_TITLE = "0.- Salir";
        public const string OPTION_1_TITLE = "1.- M. Administración";
        public const string OPTION_2_TITLE = "2.- Consulta de libros";
        public const string OPTION_3_TITLE = "3.- Préstamo de libros";
        public const string OPTION_4_TITLE = "4.- Devolución de libros";

        private GestorBibliotecaService gestorBiblioteca;

        public MainMenu(GestorBibliotecaService gestorBiblioteca)
        {
            this.gestorBiblioteca = gestorBiblioteca;
        }

        public void showMainMenu()
        {
            Console.WriteLine(MAIN_MENU_TITLE);
            Console.WriteLine("*******************************");
            Console.WriteLine(OPTION_1_TITLE);
            Console.WriteLine(OPTION_2_TITLE);
            Console.WriteLine(OPTION_3_TITLE);
            Console.WriteLine(OPTION_4_TITLE);
            Console.WriteLine(OPTION_0_TITLE);
        }

        public void executeOption0()
        {
            System.Environment.Exit(0);
        }

        public void executeOption1()
        {
            try
            {
                String password = "";

                Console.WriteLine("Por favor inserte la contraseña de administración:");
                password = Console.ReadLine();

                int connectionResult = gestorBiblioteca.Conexion(password);

                switch (connectionResult)
                {
                    case -1:
                        Console.WriteLine("ERROR: Ya hay un usuario identificado como administrador");
                        break;
                    case -2:
                        Console.WriteLine("ERROR: Contraseña errónea");
                        break;
                    default:
                        GestorBibliotecaUserProperties.getInstance().AdminId = connectionResult;

                        AdministrationMenu adminMenu = new AdministrationMenu(gestorBiblioteca);

                        int option = -1;

                        do
                        {
                            adminMenu.show();

                            Console.WriteLine("Selecciona una opción:");
                            Int32.TryParse(Console.ReadLine(), out option);

                            adminMenu.executeOption(option);

                        } while (option != 0);

                        break;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error de conexión: " + ex.ToString());
            }
        }

        public void executeOption2()
        {
            string searchText = "";

            Console.WriteLine("Introduce el texto a buscar: ");
            searchText = Console.ReadLine();

            string searchCode = "";

            Console.WriteLine("Código de Búsqueda");
            Console.WriteLine("I.- Por ISBN");
            Console.WriteLine("T.- Por Título");
            Console.WriteLine("A.- Por Autor");
            Console.WriteLine("P.- Por País");
            Console.WriteLine("D.- Por Idioma");
            Console.WriteLine("*.- Por todos los campos");

            do
            {
                Console.WriteLine("Introduce código: ");
                searchCode = Console.ReadLine();

                if (searchCode.ToLowerInvariant() != "i" && searchCode.ToLowerInvariant() != "t"
                        && searchCode.ToLowerInvariant() != "a" && searchCode.ToLowerInvariant() != "p"
                        && searchCode.ToLowerInvariant() != "d" && searchCode != "*")
                {
                    Console.WriteLine("ERORR: El código introducido no es válido");
                }
            } while (searchCode.ToLowerInvariant() != "i" && searchCode.ToLowerInvariant() != "t"
                        && searchCode.ToLowerInvariant() != "a" && searchCode.ToLowerInvariant() != "p"
                        && searchCode.ToLowerInvariant() != "d" && searchCode != "*");

            try
            {
                BookUtils bookUtils = new BookUtils();

                int adminId = GestorBibliotecaUserProperties.getInstance().AdminId;

                int totalNumberOfBooks = gestorBiblioteca.NLibros(-1);

                if (totalNumberOfBooks <= 0)
                {
                    Console.WriteLine("ERROR: No hay repositorios cargados en la biblioteca");
                    return;
                }

                List<TLibro> allBooks = new List<TLibro>();

                for (int i = 0; i < totalNumberOfBooks; i++)
                {
                    TLibro book = gestorBiblioteca.Descargar(adminId, -1, i);

                    if (book != null)
                    {
                        allBooks.Add(book);
                    }
                }

                bool headerShow = true;

                searchCode = searchCode.ToLowerInvariant();

                if (searchCode == BookSearcher.ISBN_SEARCH_FIELD)
                {
                    for (int i = 0; i < allBooks.Count(); i++)
                    {
                        TLibro book = allBooks.ElementAt(i);

                        if (book.Isbn.Contains(searchText))
                        {
                            bookUtils.Mostrar(i, headerShow, book);
                            headerShow = false;
                        }
                    }
                }
                else if (searchCode == BookSearcher.TITLE_SEARCH_FIELD)
                {
                    for (int i = 0; i < allBooks.Count(); i++)
                    {
                        TLibro book = allBooks.ElementAt(i);

                        if (book.Titulo.Contains(searchText))
                        {
                            bookUtils.Mostrar(i, headerShow, book);
                            headerShow = false;
                        }
                    }
                }
                else if (searchCode == BookSearcher.AUTHOR_SEARCH_FIELD)
                {
                    for (int i = 0; i < allBooks.Count(); i++)
                    {
                        TLibro book = allBooks.ElementAt(i);

                        if (book.Autor.Contains(searchText))
                        {
                            bookUtils.Mostrar(i, headerShow, book);
                            headerShow = false;
                        }
                    }
                }
                else if (searchCode == BookSearcher.COUNTRY_SEARCH_FIELD)
                {
                    for (int i = 0; i < allBooks.Count(); i++)
                    {
                        TLibro book = allBooks.ElementAt(i);

                        if (book.Pais.Contains(searchText))
                        {
                            bookUtils.Mostrar(i, headerShow, book);
                            headerShow = false;
                        }
                    }
                }
                else if (searchCode == BookSearcher.LANGUAGE_SEARCH_FIELD)
                {
                    for (int i = 0; i < allBooks.Count(); i++)
                    {
                        TLibro book = allBooks.ElementAt(i);

                        if (book.Idioma.Contains(searchText))
                        {
                            bookUtils.Mostrar(i, headerShow, book);
                            headerShow = false;
                        }
                    }
                }
                else if (searchCode == BookSearcher.ALL_SEARCH_FIELD)
                {
                    for (int i = 0; i < allBooks.Count(); i++)
                    {
                        TLibro book = allBooks.ElementAt(i);

                        if (book.Isbn.Contains(searchText) || book.Titulo.Contains(searchText)
                                || book.Autor.Contains(searchText) || book.Pais.Contains(searchText)
                                || book.Idioma.Contains(searchText))
                        {
                            bookUtils.Mostrar(i, headerShow, book);
                            headerShow = false;
                        }
                    }
                }

                //This means that no book has been found meeting the criteria
                if (headerShow == true)
                {
                    Console.WriteLine("Error: No se ha encontrado ningún libro");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.ToString());
            }
        }

        public void executeOption(int optionNumber)
        {
            switch (optionNumber)
            {
                case 0:
                    executeOption0();
                    break;
                case 1:
                    executeOption1();
                    break;
                case 2:
                    executeOption2();
                    break;
                case 3:
                    //executeOption3();
                    break;
                case 4:
                    //executeOption4();
                    break;
                default:
                    Console.WriteLine("opción incorrecta");
                    break;
            }
        }
    }
}
